omnigibson/systems/macro_particle_system.py

```diff
@@ -812,92 +539,177 @@ def generate_group_particles_on_object(cls, group, max_samples, min_samples_for_
         return success
 
     @classmethod
-    def get_particles_position_orientation(cls, group=None):
+    def _compute_batch_particles_position_orientation(cls, particles, local=False):
         """
-        Computes all particles' global positions and orientations that belong to this system
-
-        Note: This is more optimized than doing a for loop with self.get_particle_position_orientation()
+        Computes all @particles' positions and orientations
 
         Args:
-            group (None or str): If specified, will only return subset of particle positions and orientations belonging
-                to @group. None results in all particles being returned
+            particles (Iterable of str): Names of particles to compute batched position orientation for
+            local (bool): Whether to compute particles' poses in local frame or not
 
         Returns:
             2-tuple:
-                - (n, 3)-array: per-particle (x,y,z) position in the world frame
-                - (n, 4)-array: per-particle (x,y,z,w) quaternion orientation in the world frame
-        """
-        # Iterate over all particles and compute link tfs programmatically, then batch the matrix transform
-        link_tfs = dict()
-        link_tfs_batch = np.zeros((cls.n_particles, 4, 4))
-        particle_local_poses_batch = np.zeros_like(link_tfs_batch)
-        particles = cls.particles if group is None else cls._group_particles[group]
-        for i, name in enumerate(particles):
-            link = cls._particles_info[name]["link"]
-            if link in link_tfs:
-                link_tf = link_tfs[link]
-            else:
-                link_tf = T.pose2mat(link.get_position_orientation())
-                link_tfs[link] = link_tf
-            link_tfs_batch[i] = link_tf
-            particle_local_poses_batch[i] = cls._particles_local_mat[name]
-
-        # Compute once
-        global_poses = np.matmul(link_tfs_batch, particle_local_poses_batch)
+                - (n, 3)-array: per-particle (x,y,z) position
+                - (n, 4)-array: per-particle (x,y,z,w) quaternion orientation
+        """
+        n_particles = len(particles)
+        if local:
+            global_poses = np.zeros((n_particles, 4, 4))
```

Comment:
`global_poses` is a bit misleading here? In general I think this mixture of global and local is extremely error-prone. Let's discuss in person to see if there is a better way.

Reply:
fixed

Reply:
Is this fixed?

Reply:
fixed